README.md
# Simplified Fetch

Encapsulate a unified API request object to simplify the use of [fetch | MDN](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) and enhance it!

[![NPM version](https://img.shields.io/npm/v/simplified-fetch?style=flat-square)](https://npmjs.org/package/simplified-fetch)
[![NPM downloads](https://img.shields.io/npm/dm/simplified-fetch?style=flat-square)](https://npmjs.org/package/simplified-fetch)

___support borwser & node.js___

> npm i simplified-fetch

__If you like this repository, share it!__

## Usage

```ts
import API, { urnParser } from 'simplified-fetch'

// generate 'Api' on globalThis/window/global(nodejs)
API.init({
    newName?: string, // default:'Api', just for global access
    baseURL?: string | URL,
    method?: Methods, // default:'GET', 'POST', 'PUT'...
    bodyMixin?: BodyMixin, // default:'json', 'text', 'blob', 'formData', 'arrayBuffer'
    enableAbort?: boolean | number, // abort & timeout
    pureResponse?: boolean, // default:false, whether resolved with Response.clone()
    // format: [response, pureResponse] or response
    suffix?: string, // like .do .json
    custom?: any, // anything you want to put inside and use it in pipeline
},{
    someApi:{
        urn: string | (params?: any) => string, // build in function: urnParser
        config?: BaseConfig, // same as the above first param
    },
    someApi2:{...},
    someApi3:'/xxx', // string as urn is supported
    someApi4: (param?: any) => string, // function as urn is also supported
})

// somewhere.js
// all params are optional
await Api.someApi(body, params, config)
// enableAbort isn't supported in dynamic config
```
```ts
// support multi instances by create
const api = API.create({...}:BaseConfig, {...}:ApiConfig)
```

>### what is BodyMixin? [body-mixin | whatwg](https://fetch.spec.whatwg.org/#body-mixin)

if you are familiar with fetch, it's just used for normal step response.json() or others.

### Example

```ts
import API from "simplified-fetch"
import type { apiF, iApi, iApi_beta, APIConfig } from "simplified-fetch"

declare global {
    // unable to hint when Api.aborts.someApiEnableAbort
    // var Api: iApi & Apis
    // able to hint when Api.aborts.someApiEnableAbort
    var Api: iApi_beta<typeof configs> & Apis
}

// type your response
type response<T> = {
    body: T,
    ok: boolean, status: number, statusText: string, type: string,
}

interface Apis {
    // you should type your own apiCallFunc, of course, you can bulid on this
    someApi0: apiF<void, void, response<{ api0: number }>>,
    someApi1: apiF<{ api1: number }, number, response<{ api1: string }>>,
    // someApi2: apiF<any, any, response<{ api2: { api2: number } }>>,
}

// comment next line after config all Apis
const configs: APIConfig<Apis> = {
// necessary to enable hint when Api.aborts.someApiEnableAbort
// const configs = {
    someApi0: '/someApi0',
    someApi1: { urn: '/someApi1', config: { method: 'GET' } },
    // someApi2: { urn: '/someApi2', config: { headers: { 'Content-Type': 'application/x-www-form-urlencoded' } } },
} as const

API.init({
    baseURL: 'https://www.example.com',
    method: 'POST',
    mode: 'cors',
}, configs)

Api.request.use((url, config) => {
    // @ts-ignore
    config.headers['Authorization'] = getToken('example')
}, () => 'No Request')

const example = async () => {
    try {
        const { body, ok } = await Api.someApi0()
        const { body: { api1 } } = await Api.someApi1({ api1: 1 }, 1)
    } catch (e) {
        console.warn(e)
    }
}
```

## Config & Ability

- ### default config & BaseConfig extends [RequestInit | whatwg](https://fetch.spec.whatwg.org/#requestinit)
BaseConfig is just an extension of second param of fetch(resource [, init]) [fetch | MDN](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch)

configs are listed in Usage.
```ts
{
  method: 'GET',
  bodyMixin: 'json',
  headers: {
    "Content-Type": "application/json",
  },
  enableAbort: false,
  pureResponse: false,
}
```

- ### urnParser & params
urnParser is based on Template strings or [Template literals | MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals)
#### usage : _executed before body formatter_

if typeof urn is function, then invike it with params, and build url with returned string.

if typeof urn isn't function, then try to transform params and append to the search of URL (for type Object, FormData, URLSearchParams), or append to the pathname of URL (for type Array, String, Number).

```ts
// init
someApi:{
  urn: urnParser`/xxx/${0}/${1}`
}
// somewhere.js
Api.someApi(body, ['user',[1,2,3]], config)
// getUrl: /xxx/user/1,2,3
```
>in a way, you can do anything dynamicly on url, just set the placeholder index on, pass an Array, even an Object (index need to be string like: ${'key'} ).

ps: if you get better idea or create some beautiful way to format the url, please PR!

- ### abort & timeout
[AbortController | MDN](https://developer.mozilla.org/en-US/docs/Web/API/AbortController)

```ts
const [constroller, signal] = Api.aborts.someApi
```
>when you use as timeout, the number is accessible on signal.timeout and (error: AbortError).timeout

- ## pipeline & control
* #### pipeline (Api.request / Api.response)
  * manage orderable functions which pipes the request or response.
  * pipes with different order executed in ascending numeric index order
  * pipes with same order executed in chronological order, like queue, first in first executed.
  * based on feature [Ordinary OwnPropertyKeys](https://262.ecma-international.org/6.0/#sec-ordinary-object-internal-methods-and-internal-slots-ownpropertykeys)


* #### use/eject with order
  * use: @param ...pipe - pipe[0] as order should be nonnegative integer, rest should be function(s) as pipe(s). if order is not specified, set order to 0, which means will be executed in the first place.
  * eject: @param pipe - pipe[0] as order should be nonnegative integer, rest should be function(s) as pipe(s). if order is not specified, set order to the value which means executed first or normally 0.
  * order range: +0 ≤ i < 2^32 - 1

* #### PipeRequest
__Asynchronous executed just before fetch and after internal core operation with url & config__
>function: (url: URL, config: BaseConfig, [body, params, dynamicConfig], [someApi, urn, config, baseConfig]) => unknown
>>__only the change to url & config will effect,__ others are just from your init/create config & call params

>Not recommended:
>Change anything in params[2 | 3] will possibly causes bugs

* #### PipeResponse
__Asynchronous executed just after get Response__
>function: (response: Response, request: Request, [resolve, reject]) => Promise<unknown\>
>>__invoke resolve | reject to end pipeline__

>Response and Request are both unique for each PipeResponse

```ts
const [order, pipes: PipeRequest[]] = Api.request.use(order: number | PipeRequest, ...functions: PipeRequest[])
// Math.abs&trunc(order), if get NaN/Infinity, may causes bugs(will executed in the last place).
// Personal Recommendations: 0b1111
const bools = Api.request.eject([order, pipes])
// remove function(s) in specific order from pipeline, return true means success

const [order, pipes: PipeResponse[]] = Api.response.use(order: number | PipeResponse, ...functions: PipeResponse[])
const bools = Api.response.eject([order, pipes])
```

- ### `control`
___PipeRequest___
function return true or any message, someApi will immediate reject with that, don't forget to catch it.

___PipeResponse___
invoke resolve | reject to end pipeline

- ### body
__Failed to execute 'fetch' on 'Window': Request with ！GET/HEAD！ method cannot have body.__

[fetch.spec.whatwg.org](https://fetch.spec.whatwg.org/#request-class) constructor step-34

so body will be auto transformed by internal function to string, append to the search of URL (for type Object, FormData, URLSearchParams), or append to the pathname of URL (for type Array, String, Number).

other methods: Object and Array will be auto wrapped by JSON.stringfy()

- ## more
__read [docs](https://benno-wu.github.io/SimplifiedFetch/) or create an issue or a discussion.__

## runtime NodeJS

- ### FormData
when using FormData, please require this [@web-std/form-data](https://github.com/web-std/io/tree/main/form-data) and set FormData global.
Don't set this [form-data](https://github.com/form-data/form-data) global, and you can still use it local.

_Reason_: When body or params type FormData, internal core operation with __url__ needs Web API compatible FormData.

# Idea & Beta

- use/eject pipe once?
- formdata better support(application/x-www-form-urlencoded | multipart/form-data)
- fake mock? PipeRequest with resolve

---
Thanks to MDN, whatwg and Many blogers...


CHANGELOG.md
### 0.6.0
#### ⚠ _notice_: big change, read bofore update.
- runtime nodejs
- add polyfill: abort-controller, isomorphic-fetch
- delete awesome globalThis polyfill by Mathias Bynens
- better typed

### 0.5.0
#### ⚠ _notice_: big change, read bofore update.
- use/eject pipes with order
- params one more usage
- dynamicConfig is accessible on pipeRequest
- add ts example
- better typed
- add eslint

### 0.4.1
**bug fix**
- body type URLSearchParams

### 0.4.0
#### ⚠ _notice_: big change, read bofore update.
- pipeRequest is async now
- support dynamic config when invoke Api.someApi(body, param, config)
- fix wrong design of param on pipeline.use
- publish iife
- much better typed
- add corejs@3
- uglify->terser, causes bug on pipeline, when test with iife.

### 0.2.0
#### ⚠ _notice_: apiF big change, read bofore update.
- add custom on BaseConfig for custom config
- add APIConfig support: someApi: string
- better typed

### 0.1.0
#### ⚠ _notice_: big change, read bofore update.
- add use/eject with array
- test passed: suffix

### 0.0.12
- better type support

### 0.0.11
- better type support

### 0.0.10
- fix polyfill export

### 0.0.9
- fix export error & forget to generate the type

known bugs
- wrong polyfill export, it always runs even if you don't import it.

### 0.0.8
- add output esm.js

now support umd & esm

### 0.0.7
- new control of pipeline
- pipeline test passed

new ideas
- use/eject with array

### 0.0.6
_notice:_ pipeline will be fixed soon and come with new design of control
- test passed: abort, pureResponse

new ideas
- design of pipe control

known bugs
* pipeline

### 0.0.3
- first vision
- base test passed

known bugs
* abort & timeout

src/API.ts
import { hasOwnProperty, objects, toString } from "./shares";
import type { APIConfig, BaseConfig, bodyAsParams, BodyMixin, iAborts, iApi, iOrderablePipe, iPipe, PipeRequest, PipeResponse, PipeUnion, request, URN, URNParser } from "./type&interface";
// tried and failed on dynamic import, hard2build
import 'isomorphic-fetch';
import "abort-controller/polyfill"
// if (process?.version > 'v15') {
//     // after NodeJS v15, native supported
// }

/**
 * main entry to generate the SimplifiedFetch
 * @public
 */
export default class API {
    /**
     * global basic config
     * @defaultValue
     * ```json
     * method: 'GET',
     * bodyMixin: 'json',
     * headers: {
     *    "Content-Type": "application/json",
     * },
     * enableAbort: false,
     * pureResponse: false,
     * ```
     */
    static readonly baseConfig: BaseConfig = {
        method: 'GET',
        bodyMixin: 'json',
        headers: {
            "Content-Type": "application/json",
        },
        enableAbort: false,
        pureResponse: false,
    }

    /**
     * init Api or [newName] on globalThis
     * @param baseConfig - {@link BaseConfig}
     * @param apis - {@link APIConfig}
     */
    static init(baseConfig: BaseConfig, apis: APIConfig<Record<string, request>>): void {
        (<any>globalThis)[baseConfig?.newName || 'Api'] = new Api(apis, mergeConfig(API.baseConfig, baseConfig))
    }
    /**
     * create and return the new Api
     * @param baseConfig - {@link BaseConfig}
     * @param apis - {@link APIConfig}
     * @returns SimplifiedFetch
     */
    static create(baseConfig: BaseConfig, apis: APIConfig<Record<string, request>>): iApi {
        return new Api(apis, mergeConfig(API.baseConfig, baseConfig))
    }
}

/**
 * The unified API request object, also 'SimplifiedFetch'
 */
class Api implements iApi {
    /**
     * for AbortController & AbortSignal
     */
    aborts: iAborts = {}
    /**
     * {@link PipeRequest}
     */
    request = new OrderablePipe<PipeRequest>()
    /**
     * {@link PipeResponse}
     */
    response = new OrderablePipe<PipeResponse>()
    /**
     * constructor of Api, return SimplifiedFetch
     * @param apis - {@link APIConfig}
     * @param baseConfig - {@link BaseConfig}
     */
    constructor(apis: APIConfig<Record<string, request>>, baseConfig: BaseConfig) {
        for (const [api, request] of Object.entries(apis)) {
            let urn: URN, config: Omit<BaseConfig, 'newName'> = {};
            if (typeof request === 'string' || typeof request === 'function') urn = request
            else { ({ urn, config = {} } = request) }

            const configMerged: BaseConfig = mergeConfig(baseConfig, config)

            let controller: AbortController, signal: AbortSignal
            const abort = configMerged?.enableAbort
            if (abort) {
                controller = new AbortController()
                signal = controller.signal
                configMerged.signal = signal;
                this.aborts[api] = [controller, signal]
            }

            (<any>this)[api] = async (body?: bodyAsParams, params?: unknown, dynamicConfig: Omit<BaseConfig, 'enableAbort' | 'newName'> = {}) => {
                // controller of both pipelines
                let isEnd: unknown = false

                const configFinal = mergeConfig(configMerged, dynamicConfig)

                const urlMerged = mergeURL(urn, configFinal, params)
                urlMerged.pathname += configFinal?.suffix ?? ''
                const urlFinal = body != null ? getURL(urlMerged, configFinal, body) : urlMerged

                // pipe request
                // too many parameters, but good for bug fix when people use this
                // for (const pipes in this.request.pipeMap)
                //     if (hasOwnProperty(this.request.pipeMap, pipes))
                //         for (const pipe of this.request.pipeMap[pipes])
                //             if (!isEnd && typeof pipe === 'function')
                //                 isEnd = await pipe(urlFinal, configFinal, [body, params, dynamicConfig], [api, urn, config, baseConfig])
                for (const pipes of this.request)
                    for (const pipe of pipes)
                        if (!isEnd && typeof pipe === 'function')
                            isEnd = await pipe(urlFinal, configFinal, [body, params, dynamicConfig], [api, urn, config, baseConfig])

                return new Promise((resolve, reject) => {

                    if (isEnd) { reject(isEnd); return }
                    const res = (_: unknown) => { isEnd = true; resolve(_) }
                    const rej = (_: unknown) => { isEnd = true; reject(_) }

                    if (typeof abort === 'number') {
                        (<any>signal)['timeout'] = abort
                        // setTimeout(controller.abort, abort)
                        // https://dev.to/bil/using-abortcontroller-with-react-hooks-and-typescript-to-cancel-window-fetch-requests-1md4
                        // abortcontroller TypeError: Illegal invocation
                        setTimeout(() => controller.abort(), abort)
                    }

                    const request = new Request(urlFinal.toString(), configFinal)
                    fetch(request.clone())
                        .then(async response => {

                            // pipe response
                            // for (const pipes in this.response.pipeMap)
                            //     if (hasOwnProperty(this.response.pipeMap, pipes))
                            //         for (const pipe of this.response.pipeMap[pipes]) {
                            //             if (typeof pipe === 'function')
                            //                 await pipe(response.clone(), request.clone(), [res, rej])
                            //             if (isEnd) return
                            //         }
                            for (const pipes of this.response)
                                for (const pipe of pipes) {
                                    if (typeof pipe === 'function')
                                        await pipe(response.clone(), request.clone(), [res, rej])
                                    if (isEnd) return
                                }

                            processResponse([resolve, reject], response, configFinal?.bodyMixin, configFinal?.pureResponse)
                        })
                        .catch(e => {
                            // https://developer.mozilla.org/en-US/docs/Web/API/DOMException
                            // if (e?.code === 20) {
                            if (e.name === 'AbortError' && typeof abort === 'number') {
                                (<any>e)['timeout'] = abort
                            }
                            reject(e)
                        })
                })
            }
        }
    }
}

/**
 * manage orderable functions which pipe the request or response,
 * based on feature [Ordinary OwnPropertyKeys](https://262.ecma-international.org/6.0/#sec-ordinary-object-internal-methods-and-internal-slots-ownpropertykeys)
 * @remarks
 * [array index | Integer indix](https://tc39.es/ecma262/multipage/ecmascript-data-types-and-values.html#array-index)
 * 
 * {@link iOrderablePipe}
 */
class OrderablePipe<T extends PipeUnion> implements iOrderablePipe<T> {
    /**
     * @remarks
     * [Dr. Axel Rauschmayer | exploringjs.com/es6](https://exploringjs.com/es6/ch_oop-besides-classes.html#_traversal-order-of-properties)
     * 
     * [stackoverflow | Y2008](https://stackoverflow.com/questions/280713/elements-order-in-a-for-in-loop)
     * 
     * [stackoverflow | Y2012](https://stackoverflow.com/questions/30076219/does-es6-introduce-a-well-defined-order-of-enumeration-for-object-properties)
     * 
     * [ES2015](https://262.ecma-international.org/6.0/#sec-ordinary-object-internal-methods-and-internal-slots-ownpropertykeys)
     * 
     * [ES2022](https://tc39.es/ecma262/multipage/ordinary-and-exotic-objects-behaviours.html#sec-ordinaryownpropertykeys)
     * 
     * [10 methods of rounding in JS](https://juejin.cn/post/6901887182375092231)
     * 
     * [performance](https://zhuanlan.zhihu.com/p/28105425)
     */
    #pipeMap: Record<number | string, T[]> = {}
    // issue about Tuple, related #4988 #17765
    // finally in [#9874](https://github.com/microsoft/TypeScript/issues/9874), label: Too Complex
    use = (...pipe: [number | string | T, ...T[]]): [number, T[]] => {
        let order: number, pipes: T[]
        // @ts-ignore 1.Math.trunc(string) 2.tuple.slice
        typeof pipe[0] !== 'function' ? (order = Math.abs(Math.trunc(pipe[0])), pipes = pipe.slice(1)) : (order = 0, pipes = pipe)
        // !Array.from(pipes)! resolve bug in beforeV0.5->APIConfig type URN test->URN type string, which will eject with length:5
        hasOwnProperty(this.#pipeMap, order) ? this.#pipeMap[order].push(...pipes) : this.#pipeMap[order] = Array.from(pipes)
        return [order, pipes]
    }
    eject = ([num, pipe]: [number | string | T, T[]]): boolean[] => {
        let order: number | string, pipes: T[]
        // @ts-ignore
        typeof num !== 'function' ? (order = Math.abs(Math.trunc(num)), pipes = pipe) : (order = Object.keys(this.pipeMap)?.[0] ?? 0, pipes = [num].concat(pipe))
        const result: boolean[] = Array(pipes.length).fill(false)
        this.#pipeMap[order] = this.#pipeMap[order].filter((v) => {
            const index = pipes.indexOf(v)
            return index === -1 ? true : (result[index] = true, false)
        })
        return result
    }
    /**
     * @internal
     */
    *[Symbol.iterator]() {
        for (const key in this.#pipeMap) {
            yield this.#pipeMap[key]
        }
    }
}

/**
 * use ordered Map to manage the pipe<function>
 * {@link iPipe}
 * @deprecated since 0.6, use OrderablePipe instead
 */
class Pipe<T> implements iPipe<T> {
    pipeMap = new Map<string, T>()
    use = (..._: T[]): string | string[] => {
        const key: string[] = []
        _.forEach(v => {
            const __ = Math.random().toString(16).slice(-3)
            this.pipeMap.set(__, v)
            key.push(__)
        })
        return key.length === 1 ? key[0] : key
    }
    eject = (key: string | string[]): boolean | boolean[] => {
        const _ = new Array<string>().concat(key)
        const result: boolean[] = []
        _.forEach(v => {
            result.push(this.pipeMap.delete(v))
        })
        return result.length === 1 ? result[0] : result
    }
}

/**
 * transform the response and resolve([pure response, res] or res)
 * @param param0 - [resolve, reject] end the pipeline
 * @param response - origin pure response, {@link https://developer.mozilla.org/en-US/docs/Web/API/Response}
 * @param bodyMixin - {@link BodyMixin}
 * @param pure - get from config, whether resolve with Response.clone()
 */
function processResponse([resolve, reject]: [(value: unknown) => void, (reason?: any) => void], response: Response, bodyMixin: BodyMixin = 'json', pure = false) {
    const pureResponse: Response | undefined = pure ? response.clone() : undefined;
    response[bodyMixin]()
        .then((res: unknown) => resolve(pure ? [res, pureResponse] : res))
    // .catch(reject)
}

/**
 * merge globalConfig with localConfig
 * @param baseConfig - {@link BaseConfig}
 * @param newConfig - {@link BaseConfig}
 * @remarks
 * shallow copy custom when it's simple object, otherwise replace. 
 * @returns new object with final config
 */
function mergeConfig(baseConfig: BaseConfig, newConfig: BaseConfig): BaseConfig {
    const headers = Object.assign({}, baseConfig?.headers, newConfig?.headers)
    return (toString(baseConfig?.custom) === objects.Object || toString(newConfig?.custom) === objects.Object) ? {
        ...baseConfig, ...newConfig, headers, custom: Object.assign({}, baseConfig?.custom, newConfig?.custom)
    } : { ...baseConfig, ...newConfig, headers, }
}

/**
 * merge all configs
 * @param configs - array of {@link BaseConfig}
 * @returns final config
 */
function mergeConfigs(...configs: BaseConfig[]) {
    return configs.reduce((pre, cur, i, a) => mergeConfig(pre, cur), {})
}

/**
 * generate fetch url
 * @param urn - {@link URN}
 * @param config - {@link BaseConfig}
 * @param params - Api.someApi(body, params, dynamicConfig), use for step: urnParser
 * @returns url wait to fetch
 */
function mergeURL(urn: URN, config: BaseConfig, params?: unknown): URL {
    const base = config?.baseURL ?? undefined
    return typeof urn === 'function' ? new URL(urn(params), base) : appendPathnameOrSearch(params, new URL(urn, base))
}

/**
 * transform body properly to fetch body
 * @param url - {@link https://developer.mozilla.org/en-US/docs/Web/API/URL}
 * @param config - {@link BaseConfig}
 * @param body - {@link bodyAsParams}
 * @returns final url and config for fetch
 */
function getURL(url: URL, config: BaseConfig, body: bodyAsParams): URL {
    // reason: Failed to execute 'fetch' on 'Window': Request with ！GET/HEAD！ method cannot have body.
    // https://fetch.spec.whatwg.org/#request-class constructor step-34
    if (['GET', 'HEAD'].includes(config.method!.toUpperCase())) {
        url = appendPathnameOrSearch(body, url)
    } else {
        //  obj2string,exclude
        //  ['[object Blob]','[object ArrayBuffer]','[object FormData]',
        //  '[object URLSearchParams]','[object ReadableStream]','[object String]']
        if (toString(body) === objects.Object || Array.isArray(body)) {
            body = JSON.stringify(body)
        }
        config.body = body as BodyInit
    }
    return url
}

/**
 * append params | body to URL pathname | search
 * @param end params or body
 * @param url URL
 * @returns URL
 * @remarks
 * type [Object, FormData, URLSearchParams] append to search
 * type [Array, String, Number] append to pathname
 */
function appendPathnameOrSearch(end: any, url: URL): URL {
    const endType = toString(end)
    // https://developer.mozilla.org/en-US/docs/Web/API/FormData
    /*
    You can also pass it directly to the URLSearchParams constructor
    if you want to generate query parameters in the way a <form> would do
    if it were using simple GET submission.
     */
    // https://github.com/microsoft/TypeScript/issues/30584
    if ([objects.Object, objects.FormData, objects.URLSearchParams].includes(endType)) {
        // https://url.spec.whatwg.org/#dom-url-search
        const search = url.search.includes('?')
        // this won't be done automatically
        url.search += search && url.search.slice(-1)[0] !== '&' ? '&' : ''
        // @ts-ignor
        url.search += new URLSearchParams(end).toString()
    } else if ([objects.Array, objects.Number, objects.String].includes(endType)) {
        const pathname = url.pathname.slice(-1)[0] === '/'
        // @ts-ignor
        url.pathname += (pathname ? '' : '/') + end.toString()
    }
    return url
}

/**
 * parse the template strings with params
 * 
 * @example
 * ```ts
 * // init|create
 * someApi:{
 *   urn: urnParser`/xxx/${0}/${1}`
 * }
 * // somewhere
 * Api.someApi(body, ['user', [1,2,3]])
 * // getUrl: /xxx/user/1,2,3
 * ```
 * @param template - template strings
 * @param placeholder - indexes of params, like $\{0\}
 * @returns {@link URNParser}
 * 
 * @remarks
 * function base on Template literals (Template strings)
 * {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals | ES6 Template literals}
 * 
 * you do can use the string type as placehloder, and give an object as params, match them with key string.
 * But the way it used would look like: urnParser`/${'key1'}/${'key2'}`, Api.someApi(body, \{key1:'',key2:''\})
 * 
 * Anyway, need better idea.
 * 
 * @public
 */
export const urnParser = (template: Array<string>, ...placeholder: Array<number | string>): URNParser => {
    return params => template.reduce((previousValue, currentValue, index) => {
        // https://github.com/microsoft/TypeScript/issues/10530
        return previousValue + currentValue + (params?.[placeholder[index]] ?? '')
    }, '')
}
